src/internal_modules/roc_core/mov_histogram.h

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2023 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/mov_histogram.h
+//! @brief Rolling window moving histogram.
+
+#ifndef ROC_CORE_MOV_HISTOGRAM_H_
+#define ROC_CORE_MOV_HISTOGRAM_H_
+
+#include "roc_core/array.h"
+#include "roc_core/iarena.h"
+#include "roc_core/panic.h"
+#include "roc_core/ring_queue.h"
+
+namespace roc {
+namespace core {
+
+/// @brief
+//! Efficiently implements moving histogram based on approach
+/// @tparam T
+template <typename T> class MovHistogram {
+public:
+    //! Initialize.
+    MovHistogram(IArena& arena, T value_range, size_t num_bins, size_t window_length)
+        : value_range_(value_range)
+        , num_bins_(num_bins)
+        , window_length_(window_length)
+        , bin_width(static_cast<T>(value_range) / static_cast<T>(num_bins))
+        , ring_buffer_(arena, window_length)
+        , bins_(arena)
+        , valid_(false) {
+        if (num_bins == 0 || window_length_ == 0) {
+            roc_panic(
+                "mov histogram: number of bins and window length must be greater than 0");
+        }
+
+        if (!ring_buffer_.is_valid() || !bins_.resize(num_bins)) {
+            return;
+        }
+
+        valid_ = true;
+    }
+
+    //! Check if the histogram is valid.
+    bool is_valid() const {
+        return valid_;
+    }
+
+    //! Add a value to the histogram.
+    void add_value(const T& value) {
+        if (ring_buffer_.size() == window_length_) {
+            T oldest_bin_ = ring_buffer_.front();
+            ring_buffer_.pop_front();
+            int oldest_bin_index = getBinIndex(oldest_bin_);
+            if (oldest_bin_index >= 0
+                && static_cast<size_t>(oldest_bin_index) < num_bins_) {
+                bins_[static_cast<size_t>(oldest_bin_index)]--;
+            }
+        }
+
+        ring_buffer_.push_back(value);
+        int new_bin_index = getBinIndex(value);
+        if (new_bin_index >= 0 && static_cast<size_t>(new_bin_index) < num_bins_) {
+            bins_[static_cast<size_t>(new_bin_index)]++;
+        }
+    }
+
+    //! returns the number of values in the bin
+    int getBinCounter(size_t bin_index) const {
+        return bins_[bin_index];
+    }
+
+private:
+    //! returns the bin index for a given value
+    int getBinIndex(const T& value) const {
+        if (value > value_range_) {
+            return -1;
```

Comment:
Consider validating `value` in `add_value()` instead of using a placeholder here. Therefore the returning type will be size_t.

I asked Victor, he told me you can use saturation there:
- if value > max => value = max
- if value < min => value = min